File: patterns/go-security-audit/test-vulnerable.go
package main

import (
	"crypto/md5"
	"database/sql"
	"fmt"
	"io/ioutil"
	"math/rand"
	"net/http"
	"os/exec"
	"text/template"
	"time"
)

// VULNERABILITY: Hardcoded credentials
const (
	dbPassword = "SuperSecret123!"
	apiKey     = "sk-1234567890abcdef"
)

// VULNERABILITY: SQL Injection
func getUserByID(db *sql.DB, userID string) error {
	query := fmt.Sprintf("SELECT * FROM users WHERE id='%s'", userID)
	rows, err := db.Query(query)
	if err != nil {
		return err
	}
	defer rows.Close()
	return nil
}

// VULNERABILITY: Command Injection
func runCommand(userInput string) error {
	cmd := exec.Command("/bin/sh", "-c", "ls "+userInput)
	output, err := cmd.Output()
	if err != nil {
		return err
	}
	fmt.Println(string(output))
	return nil
}

// VULNERABILITY: Path Traversal
func readFile(filename string) ([]byte, error) {
	filepath := "/var/www/uploads/" + filename
	return ioutil.ReadFile(filepath)
}

// VULNERABILITY: XSS using text/template
func renderTemplate(w http.ResponseWriter, userContent string) error {
	tmpl := template.Must(template.New("page").Parse(`
		<html>
		<body>
			<div>{{.Content}}</div>
		</body>
		</html>
	`))
	return tmpl.Execute(w, map[string]string{"Content": userContent})
}

// VULNERABILITY: Weak cryptography (MD5)
func hashPassword(password string) string {
	hasher := md5.New()
	hasher.Write([]byte(password))
	return fmt.Sprintf("%x", hasher.Sum(nil))
}

// VULNERABILITY: Insecure random number generation
func generateToken() string {
	rand.Seed(time.Now().UnixNano())
	return fmt.Sprintf("%d", rand.Int())
}

// VULNERABILITY: Race condition
var counter int

func incrementCounter() {
	counter++
}

// VULNERABILITY: Missing TLS, no authentication
func startServer() {
	http.HandleFunc("/user", func(w http.ResponseWriter, r *http.Request) {
		userID := r.URL.Query().Get("id")
		getUserByID(nil, userID)
	})

	http.HandleFunc("/execute", func(w http.ResponseWriter, r *http.Request) {
		cmd := r.URL.Query().Get("cmd")
		runCommand(cmd)
	})

	http.HandleFunc("/file", func(w http.ResponseWriter, r *http.Request) {
		filename := r.URL.Query().Get("name")
		content, _ := readFile(filename)
		w.Write(content)
	})

	// Missing TLS
	http.ListenAndServe(":8080", nil)
}

func main() {
	// Launch goroutines without synchronization
	for i := 0; i < 100; i++ {
		go incrementCounter()
	}

	startServer()
}

File: patterns/python-security-audit/test-vulnerable.py
"""
Deliberately vulnerable Python code for testing the python-security-audit pattern.
DO NOT use this code in production.
"""

import hashlib
import os
import pickle
import random
import sqlite3
import subprocess
import yaml
from flask import Flask, request, render_template_string
from markupsafe import Markup

app = Flask(__name__)

# VULNERABILITY: Hardcoded credentials
app.secret_key = "dev-secret-key-hardcoded"
DB_PASSWORD = "SuperSecret123!"
API_KEY = "sk-1234567890abcdef"
JWT_SECRET = "my-hardcoded-jwt-secret"

# VULNERABILITY: SQL Injection
def get_user_by_id(user_id: str):
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    query = f"SELECT * FROM users WHERE id = '{user_id}'"
    cursor.execute(query)
    return cursor.fetchone()

# VULNERABILITY: SQL Injection via % formatting
def search_users(name: str):
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    query = "SELECT * FROM users WHERE name = '%s'" % name
    cursor.execute(query)
    return cursor.fetchall()

# VULNERABILITY: Command Injection
def run_ping(host: str):
    os.system(f"ping -c 4 {host}")

# VULNERABILITY: Command Injection via subprocess shell=True
def list_files(directory: str):
    result = subprocess.run(f"ls {directory}", shell=True, capture_output=True, text=True)
    return result.stdout

# VULNERABILITY: Path Traversal
def read_file(filename: str) -> str:
    path = "/var/www/uploads/" + filename
    with open(path) as f:
        return f.read()

# VULNERABILITY: Insecure deserialization (pickle)
def load_session(data: bytes):
    return pickle.loads(data)

# VULNERABILITY: Unsafe YAML loading
def load_config(config_text: str):
    return yaml.load(config_text)

# VULNERABILITY: eval() with user input
def calculate(expression: str):
    return eval(expression)

# VULNERABILITY: Weak password hashing (MD5)
def hash_password(password: str) -> str:
    return hashlib.md5(password.encode()).hexdigest()

# VULNERABILITY: SHA-1 without salt
def legacy_hash(password: str) -> str:
    return hashlib.sha1(password.encode()).hexdigest()

# VULNERABILITY: Insecure random number generation
def generate_token() -> str:
    return str(random.randint(100000, 999999))

def generate_session_id() -> str:
    return "".join(random.choices("abcdefghijklmnopqrstuvwxyz0123456789", k=32))

# VULNERABILITY: XSS via Markup
@app.route("/greet")
def greet():
    name = request.args.get("name", "")
    return Markup(f"<h1>Hello {name}</h1>")

# VULNERABILITY: Server-Side Template Injection (SSTI)
@app.route("/template")
def ssti_endpoint():
    name = request.args.get("name", "")
    template = f"<h1>Hello {name}!</h1>"
    return render_template_string(template)

# VULNERABILITY: Missing TLS verification
def fetch_data(url: str):
    import requests
    return requests.get(url, verify=False).json()

# VULNERABILITY: Debug mode enabled
@app.route("/debug")
def debug_info():
    import sys
    # Exposes internal state
    return str(sys.path) + str(os.environ)

# VULNERABILITY: Missing authentication on sensitive endpoint
@app.route("/admin/users")
def list_all_users():
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM users")
    return str(cursor.fetchall())

# VULNERABILITY: Logging sensitive data
import logging
logging.basicConfig(level=logging.DEBUG)

def authenticate(username: str, password: str) -> bool:
    logging.info(f"Login attempt: username={username} password={password}")
    return True

# VULNERABILITY: SSRF - user-controlled URL
@app.route("/proxy")
def proxy():
    import requests
    url = request.args.get("url")
    return requests.get(url).text

if __name__ == "__main__":
    # VULNERABILITY: debug=True in production
    app.run(debug=True, host="0.0.0.0", port=5000)
